File: 백준/Gold/5904. Moo 게임/Moo 게임.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        long n = Long.parseLong(br.readLine());

        int k = 0;
        long len = 3;

        while (len < n) {
            k++;
            len = len * 2 + (k + 3);
        }

        System.out.println(solve(n, k, len));
    }

    private static char solve(long n, int k, long len) {
        if (k == 0) {
            return "moo".charAt((int) (n - 1));
        }

        long prevLen = (len - (k + 3)) / 2;
        long ms = prevLen + 1;
        long me = prevLen + (k + 3);

        if (n <= prevLen) {
            return solve(n, k - 1, prevLen);
        } else if (ms <= n && n <= me) {
            return (n == ms) ? 'm' : 'o';
        } else {
            return solve(n - me, k - 1, prevLen);
        }
    }
}

File: 백준/Gold/5904. Moo 게임/Moo 게임.py
import sys


def solve(n, k, length):
    if k == 0:
        return 'm' if n == 1 else 'o'

    prev_len = (length - (k + 3)) // 2
    ms = prev_len + 1
    me = prev_len + (k + 3)

    if n <= prev_len:
        return solve(n, k - 1, prev_len)
    elif ms <= n <= me:
        return 'm' if n == ms else 'o'
    else:
        return solve(n - me, k - 1, prev_len)


n = int(input())

k, length = 0, 3

while length < n:
    k += 1
    length = length * 2 + (k + 3)

print(solve(n, k, length))
